Examples HLD (#827)
ni/nimble: specs/examples/README.md

> # Nimble Examples

Author: I'm gonna close this PR for now. Some things I think would be useful if re-opening:

- Split out the new cdn / bundle packages as separate work
- Consider completely narrowing the scope of the examples to just getting started examples. At which point maybe they should just create template packages:
  - `dotnet new install`: https://learn.microsoft.com/en-us/dotnet/core/tools/dotnet-new-install
  - `npm init`: https://docs.npmjs.com/cli/v10/commands/npm-init
  - Pros: folks don't have to sync down the repo to get started / figure out how to download just a specific folder (which GitHub doesn't make easy)

  - Blazor WASM examples will not have the routing restrictions we have for the all-components examples. This will make the URLs to the sub routes less useful if copy and pasted but allow better examples demonstrating routing.
  - While additional routes in examples are not prevented, they should be discouraged except for the use-case of demonstrating routing. It's likely extra routes make for a less coherent and targeted example anyway.

### Migrate CDN Builds to a separate package

Part of this proposal is to migrate the prebuilt all-components bundles from nimble-components to a separate npm package.

This has a couple benefits:
- Avoids confusion within the Nimble Components package. Currently nothing prevents a user from accidentally mixing usage of the normal nimble component es modules, the all-components.js es module, or the all-components-bundle.js built output. 

  Mixing them is very likely to break the application in hard to predict ways by creating multiple singletons in global scope, duplicating element registrations, etc. Moving them all to a separate package can eliminate possible confusion.

> - Can inline all the related content, i.e. the needed nimble fonts can be copied from nimble-tokens into the cdn package making the package standalone. Doing that with the nimble-components package would not be appropriate.

Reviewer: Some questions that might help me understand the vision for this package, which could help with naming and with deciding what is/isn't part of it.
1. would apps ever have an npm dependency on it or just a runtime dependency on the build hosted on unpkg?
2. would we publish a Nuget equivalent or just an npm pakcage?
3. a client could use either `nimble-bundle` or a subset of the other packages, but never mix them, correct? So any assets an app might need will need to be a part of it: fonts, icons, CSS, tokens, components?

> - Proposed package name `@ni/nimble-bundle`

Reviewer 2: I don't have a strong opinion, but I'd lean towards the alternative of `@ni/nimble-components-bundle` to make the package name better convey what is included. This isn't a bundle that includes all of nimble (Angular support + web component support + any/all other components we add first-class support for in the future).

    ```
    - wwwroot/
      - nimble-components/all-components-bundle.min.js
      - nimble-tokens/css/fonts.css
      - nimble-tokens/assets/fonts.css
    ```
    to the following where dist is a direct copy of the nimble-bundle `dist` output:
    ```
    - wwwroot/
      - dist/all-components-bundle.min.js
      - dist/css/fonts.css
      - dist/assets/fonts.css
    ```
  - Alternatives:
    - Don't introduce a breaking change, keep the existing `wwwroot` file structure in Nimble Blazor
      - Pros: Not a breaking change
      - Cons: Nimble Blazor continues to maintain an alternate structure of bundled output

### File structure

Examples will be organized as follows:

```
- examples/
    - README.md describing folder layout and documenting example expectations (linting, etc).
    - <framework>/
        - README.md: Table of contents for the examples
        - getting-started/
            - README.md: Should include a Stackblitz link to open the example immediately in addition to other content
            - package.json
            - source/
        - <other examples>: TBD
```

Notes:
- A subfolder per `<framework>`, i.e.: `vanilla`, `cdn`, `angular`, `blazor`

> - Examples should be specific and targeted

Reviewer: Will each example be totally standalone or would we have an Angular workspace containing an application per example?

  - Large examples or examples with multiple routes should be discouraged (but are not prohibited as to allow for examples demonstrating routing).
- Each example participates in the monorepo build

> - Each example follows Web Chapter conventions:

Reviewer: ```suggestion
- Each example follows Web UI Chapter conventions:
```

    - Note: This is a breaking URL change. For example, `https://nimble.ni.dev/storybook/example-client-app/` will change to `https://nimble.ni.dev/storybook/all-components-angular/`

Alternatives:
- Move the all-components examples to the examples/ folder
    - Would hurt local development experience, i.e. need to run multiple angular builds / watch as a separate package not managed by angular-workspace
    - Does not seem like a high value user facing example. The examples are very large and not illustrative of specific concepts.
    - Users should rely on storybook for per control usage instead
- Don't rename the current client examples
  - Pros: Doesn't break existing URLs
  - Cons: The current names are ambiguous about framework, i.e. `example-client-app` vs `blazor-client-app`. They also don't reflect the modern use-case of being a simple dev focused example showing all the components being used.

## Alternative Implementations / Designs

N/A

## Open Issues

### Publish examples to GitHub Pages

Should we publish the examples to GitHub pages as well as make them accessible from stackblitz.

Pros:
- Easier to open on mobile devices

> - More reliable manual stress test environment (don't have to consider additional injected page content)

Reviewer: >(don't have to consider additional injected page content)

Could you clarify what this means? Are you referring to stuff that Stackblitz adds to the page?

I have sometimes been annoyed when opening Stackblitz examples when they take more than a couple seconds to spin up their container. Having examples that avoid that would be nice. Obviously we have to consider the cost to implement it, but I like keeping it as part of the proposal.

> # Nimble Examples

Author: Something to maybe be inspired by: https://examples.deno.land/

This HLD sketches out the organization and requirements of new examples and of the current examples.

## Links To Relevant Work Items and Reference Material

- [#726 Create a "Getting Started" walkthrough for Nimble Angular](https://github.com/ni/nimble/issues/726)
- [#325 Set up stackblitz or codepen template with Nimble](https://github.com/ni/nimble/issues/325)

## Implementation / Design

### Frameworks to support
- Vanilla: TypeScript + SCSS
    - No explicit examples supporting vanilla JavaScript + CSS with local npm builds. Strongly encourage TS + SCSS.
    - Examples will use [vite](https://vitejs.dev/)
        - Supports TS + SCSS out of the box
        - Web Chapter does not have an alternate recommendation yet for Vanilla apps
        - Vite works on Stackblitz and local npm builds
            - Parcel [does not work on Stackblitz](https://github.com/parcel-bundler/parcel/issues/7748)
            - Rollup requires extra config for TS and SCSS
            - Webpack aligns with the underlying technology of Angular apps but adds lots of complexity to self manage (Angular generally manages for you)

> - CDN: Vanilla JS and CSS

Author: Note to self, using vite to build a library isn't great as it inlines a bunch of assets into JS: https://github.com/vitejs/vite/issues/4454

It's mainly good for the final app build

Author: dart-sass with a custom resolver may be an alternative
can maybe use https://github.com/wessberg/sass-extended-importer

    - Uses the CDN builds of packages via unpkg.com
    - Known issue: fonts are not included in the [existing example](https://github.com/ni/nimble/tree/0fe8cc54f7bd444830e3974969519593b532b083/packages/nimble-components#prototyping-in-a-static-webpage)
        - We should move all the pre-built cdn bundle stuff to a [separate package](https://twitter.com/justinfagnani/status/1521953590931189760?s=20&t=zejlKVWCbEBce8zYFO7Uyg), ie:
            - `@ni/nimble-bundle`
                - alternative names:
                    - `@ni/nimble-components-cdn`
                    - `@ni/nimbl-components-bundle`
                    - `@ni/nimble-cdn-bundle`
                    - `@ni/nimble-cdn`
        - Alternative:
            - We could add a build of our CSS + font assets to the existing nimble-components bundle build which will grow the package size and pose the risk of none cdn users accidentally leveraging the contents (as described in the separate package thread linked above).
- Angular
- Blazor

### File structure
  - examples/

>     - `<framework>`/

Author: One concern to investigate is the build performance cost of have potentially many small Angular examples. We should verify that the base cost of compiling multiple small examples does not explode build times.

Reviewer 3: We could also consider not having the examples be built as part of the CI build, particularly if they won't be linked off of Storybook.

Author: Not as concerned about the CI build as local build. If anything might exclude it from local build but have it run on CI.

Author: Another option is to setup one of the task orchestrators discussed in https://github.com/ni/nimble/issues/376

Author: Captured as open issue

        - README.md: Table of contents for the examples
        - getting-started/
            - README.md: Should include a Stackblitz link to open the example immediately in addition to other content
            - package.json
            - source/

>         - `<other examples>`: TBD

Reviewer: If we're not including the existing kitchen sink examples here, I'm trying to imagine a few possible examples that we would include. Just trying to picture the future to see how it might influence this structure. Some ideas:
1. pared-down real-world apps. e.g. a login screen, a master-detail pattern, a header bar with navigation menu, a basic form, a tile data visualization dashboard
2. richer component-specific examples. e.g. show how to handle a specific event, bind to a specific property, provide rich custom content.
3. stress tests. e.g. lots of copies of components or lots of data in components or rapid updates to components
4. proof of concept for frameworks that we don't support first class. e.g. how to use nimble-components in a React app or VSCode plugin or browser extension or WebVI

> 

Reviewer: Should we deploy all the built examples to GitHub Pages in addition to being able to run them in Stackblitz? I think that could be useful for doing things like running performance tests on stress test examples and for sending people direct links to demos without the chrome of stackblitz (e.g. for a PO who wants to see what the nimble-table can do but doesn't care what its API is).

If we do this, we should ensure each copy of an example crosslinks to all the other copies: readme <-> hosted example <-> example source code <-> stackblitz <-> published npm package

Reviewer: If we do this we might also want to host a Table of Contents site for examples and link to it from nimble.ni.dev.

Author: I think I initially lean against that. Our GitHub pages build is very large, ~100MB, due to storybook metadata being included and the bulkiness of individual blazor builds. There is a lot of extra bulk because these are individual app builds so framework overhead is duplicated (vs monorepo where duplicate framework cost is amortized).

> running performance tests on stress test examples 

I don't think if we had automated perf tests we would point to the published GitHub pages example. I think it would be part of the CI pipeline / an npm script you can run at any point by checking out the repo and running perf tests. Testing GitHub pages limits you to just the latest published examples.

Some additional reasons to do it:

- Easier to pull up examples in a mobile environment (i.e. manually open table stress test on phone)
- Nice shareable example URLs

I'll capture it in future work since I think It'll need some research. Maybe stripping out some stuff from the GitHub pages build, etc.

Notes:
- A subfolder per `<framework>`, i.e.: `vanilla`, `cdn`, `angular`, `blazor`
- Each `<framework>` will have a `getting-started` example.
    - Getting started example will be linked from each library package README
    - Getting started examples assumes brand new developers who may not have web development experience:
        - guidance on IDEs
        - step-by-step instructions

>         - links to learn more about general web topics they need to be aware of (CSS custon properties, shadow slots, etc.)

Reviewer 2: ```suggestion
        - links to learn more about general web topics they need to be aware of (CSS custom properties, shadow slots, etc.)
```

- Each example participates in the monorepo build
- Each example follows Web Chapter conventions:
    - Linting configured

>     - Testing configured

Author: I was on the fence about including testing in the examples but should we go even further and include storybook (for vanilla and chromatic)?

Reviewer 3: I guess I'm not quite following. Is the suggestion that each example will have tests that accompany it, and you're also asking if each example should have a storybook? I'm not sure what that means. Do you mean should the examples be part of the Chromatic tests (which happen via Storybook presence)?

If my understanding is correct, I would vote "no" on both accounts. While I think the extra coverage in Chromatic tests would be nice, it still feels a bit odd, and I'm not even sure how that would work. Testing the examples just seems like overkill.

Author: The examples are each standalone projects. The idea I was going for is that you can download any of the examples, rename it, but it up for code review and it would have everything you would expect for NI best practices for applications (linting, testing, etc).

So it's not that the examples are tested, it's that the examples have the correct test architecture set-up for someone who is getting started.

Reviewer 3: Given that I think the Storybook aspect might be too much. It's somewhat imposing our will beyond a proper boundary. I think it's sufficient for us to have a doc that could provide the guidance to a user on how to integrate Storybook into their own project, if that was easy enough to document.

Reviewer: I lean towards not requiring that each example include tests. We could optionally include them for certain examples (particularly getting started or common application patterns) but they might be a distraction that doesn't add much value for more trivial examples.

Author: Yea it does seem unnecessary for all examples. I'll update to say just for the getting started example and just unit tests (not more opionated / env specific like integration / chromatic)

    - Detailed README
- Issue templates should be updated to include the Stackblitz urls to make it easier to report issues

> ### Exisiting examples

Reviewer 2: ```suggestion
### Existing examples
```

- The current [angular client example](https://github.com/ni/nimble/tree/main/angular-workspace/projects/example-client-app) and [blazor client example](https://github.com/ni/nimble/tree/main/packages/nimble-blazor/Examples)
    - Will stay in current locations (angular-workspace and blazor workspace) as they are primarily developer oriented tools and not user facing examples
    - Their READMEs should direct end users to the `<framework>` examples folders

>     - They will be renamed to `kitchen-sink` examples
>     - Continue to not have a Vanilla kitchen sink example
>     - Alternatives:
>         - Move the kitchen-sink examples to the examples/ folder

Reviewer: After [discussion](https://github.com/ni/nimble/pull/1107#discussion_r1130170961) in a PR changing the title of these examples I vote for a name like `all-components` rather than `kitchen-sink`. It's not a showstopper, but idioms like "kitchen sink" are a bit less approachable for people whose first language language isn't English (not all languages use that idiom).

```suggestion
    - They will be renamed to `all-components` examples
    - Continue to not have a Vanilla all components example
    - Alternatives:
        - Move the all-components examples to the examples/ folder
```

            - Would hurt local development experience, i.e. need to run multiple angular builds / watch as a separate packge not managed by angular-workspace
            - Does not seem like a high value user facing example
            - Users should rely on storybook for per control usage instead

> 

Author: We should add table performance examples to experiment to play with manually.

Author: Covered by the example type discussion section

- The [counter example](https://github.com/ni/nimble/pull/796)
    - Will reformat the example to be the `getting-started` example for vanilla `<framework>`, i.e. `/examples/vanilla/getting-started/<example files>`
    - Will only include the final part of the example

>     - README will be updated to behave as a standalone example instead of a guided help / codelabs style example

Reviewer 4: Where does the instructional information go? It needs to exist somewhere.
